pkg/security/apiserver/admission/sccadmission/admission.go

```diff
@@ -71,35 +71,67 @@ func NewConstraint() *constraint {
 // we get a true old/new set of objects in.
 func (c *constraint) Admit(a admission.Attributes) error {
-	if a.GetResource().GroupResource() != kapi.Resource("pods") {
+	if ignore, err := shouldIgnore(a); err != nil {
+		return err
+	} else if ignore {
 		return nil
 	}
-	if len(a.GetSubresource()) != 0 {
+	pod := a.GetObject().(*kapi.Pod)
+
+	// TODO(liggitt): allow spec mutation during initializing updates?
+	specMutationAllowed := a.GetOperation() == admission.Create
+
+	allowedPod, sccName, validationErrs, err := c.computeSecurityContext(a, pod, specMutationAllowed, "")
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+
+	if allowedPod != nil {
+		*pod = *allowedPod
+		// annotate and accept the pod
+		glog.V(4).Infof("pod %s (generate: %s) validated against provider %s", pod.Name, pod.GenerateName, sccName)
+		if pod.ObjectMeta.Annotations == nil {
+			pod.ObjectMeta.Annotations = map[string]string{}
+		}
+		pod.ObjectMeta.Annotations[allocator.ValidatedSCCAnnotation] = sccName
 		return nil
 	}
 
-	pod, ok := a.GetObject().(*kapi.Pod)
-	// if we can't convert then fail closed since we've already checked that this is supposed to be a pod object.
-	// this shouldn't normally happen during admission but could happen if an integrator passes a versioned
-	// pod object rather than an internal object.
-	if !ok {
-		return admission.NewForbidden(a, fmt.Errorf("object was marked as kind pod but was unable to be converted: %v", a.GetObject()))
+	// we didn't validate against any security context constraint provider, reject the pod and give the errors for each attempt
+	glog.V(4).Infof("unable to validate pod %s (generate: %s) against any security context constraint: %v", pod.Name, pod.GenerateName, validationErrs)
+	return admission.NewForbidden(a, fmt.Errorf("unable to validate against any security context constraint: %v", validationErrs))
+}
+
+func (c *constraint) Validate(a admission.Attributes) error {
+	if ignore, err := shouldIgnore(a); err != nil {
+		return err
+	} else if ignore {
+		return nil
 	}
+	pod := a.GetObject().(*kapi.Pod)
 
-	// if this is an update, see if we are only updating the ownerRef.  Garbage collection does this
-	// and we should allow it in general, since you had the power to update and the power to delete.
-	// The worst that happens is that you delete something, but you aren't controlling the privileged object itself
-	if a.GetOldObject() != nil && rbacregistry.IsOnlyMutatingGCFields(a.GetObject(), a.GetOldObject(), kapihelper.Semantic) {
+	// compute the context. Mutation is not allowed. ValidatedSCCAnnotation is used as a hint to gain same speed-up.
+	allowedPod, _, validationErrs, err := c.computeSecurityContext(a, pod, false, pod.ObjectMeta.Annotations[allocator.ValidatedSCCAnnotation])
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+	if apiequality.Semantic.DeepEqual(pod, allowedPod) {
```

Comment:
need to make sure allowedPod is non-nil here

Reply:
> need to make sure allowedPod is non-nil here

Just in case a deep equals of nil matches an actual pod?